### PR TITLE
Handle if project is undetected

### LINKED ISSUE #2961: Nil pointer reference on azd up (Buildpacks)
I was recently working on https://github.com/savannahostrowski/savannah.dev/tree/main and accidentally deleted my Dockerfile. When I ran `azd up`, I got this error. I'm assuming it's because Buildpacks don't have a Go target so it just sort of freaked out? Once I added the Dockerfile back, things were all good.

That said, we should probably catch this and throw something more human-readable.

```
azd deploy --debug
2023/11/12 03:42:58 main.go:54: azd version: 1.4.4 (commit 45a48a32c9f3caad481343869ade05221df26ec7)
2023/11/12 03:42:58 project.go:113: Reading project from file '/workspaces/terminal-personal-site/azure.yaml'
2023/11/12 03:42:58 main.go:208: using cached latest version: 1.4.3 (expires on: 2023-11-12T22:13:12Z)
2023/11/12 03:42:58 cobra_builder.go:141: Resolved action 'azd-deploy-action'
2023/11/12 03:42:58 middleware.go:124: running middleware 'debug'
2023/11/12 03:42:58 middleware.go:124: running middleware 'experimentation'
2023/11/12 03:42:58 experimentation.go:42: assignment context: initflow_t:78271;0g5ad841:76970;
2023/11/12 03:42:58 middleware.go:124: running middleware 'telemetry'
2023/11/12 03:42:58 telemetry.go:50: TraceID: c2d91e82452420b886ebbb4af86ba635
2023/11/12 03:42:58 middleware.go:124: running middleware 'hooks'
2023/11/12 03:42:58 hooks.go:156: service 'web' does not require any command hooks.
2023/11/12 03:42:58 hooks.go:85: azd project is not available or does not contain any command hooks, skipping command hook registrations.
2023/11/12 03:42:58 command_runner.go:307: Run exec: 'docker --version' , exit code: 0
-------------------------------------stdout-------------------------------------------
Docker version 24.0.7-1, build afdd53b4e341be38d2056a42113b938559bb1d94
2023/11/12 03:42:58 docker.go:254: docker version: Docker version 24.0.7-1, build afdd53b4e341be38d2056a42113b938559bb1d94
2023/11/12 03:42:58 docker.go:185: determining version from docker --version string: Docker version 24.0.7-1, build afdd53b4e341be38d2056a42113b938559bb1d94
2023/11/12 03:42:58 docker.go:197: extracted docker version: 24.0.7-1, build: afdd53b4e341be38d2056a42113b938559bb1d94 from version string
2023/11/12 03:42:58 ensure.go:47: Skipping install check for 'Docker'. It was previously confirmed.

Deploying services (azd deploy)

  |       | Deploying service web2023/11/12 03:42:58 framework_service_docker.go:157: building image for service web, cwd: /workspaces/terminal-personal-site, path: ./Dockerfile, context: ., buildArgs: [])
  |      =| Deploying service web (Building Docker image from source)2023/11/12 03:42:58 command_runner.go:307: Run exec: '/home/vscode/.azd/bin/pack --version' , exit code: 0
-------------------------------------stdout-------------------------------------------
0.30.0+git-c38f7da.build-4952
2023/11/12 03:42:58 pack.go:130: pack version: 0.30.0
2023/11/12 03:42:58 pack.go:144: using local pack: /home/vscode/.azd/bin/pack
2023/11/12 03:42:58 appdetect.go:246: Detecting projects in directory: /workspaces/terminal-personal-site
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5608439de60d]

goroutine 32 [running]:
[github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).packBuild(0xc00047ff80,](http://github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).packBuild(0xc00047ff80,) {0x56084443eb48, 0xc000332e70}, 0xc000368ea0, {{0x560843cc080c, 0xc}, {0x560843e55538, 0x1}, {0x560843cbf56b, 0xb}, ...}, ...)
        /mnt/vss/_work/1/s/cli/azd/pkg/project/framework_service_docker.go:323 +0x46d
[github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).Build.func1(0xc0000bb920)](http://github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).Build.func1(0xc0000bb920))
        /mnt/vss/_work/1/s/cli/azd/pkg/project/framework_service_docker.go:182 +0x748
[github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run.func1()](http://github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run.func1())
        /mnt/vss/_work/1/s/cli/azd/pkg/async/task.go:144 +0xfd
created by [github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run](http://github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run) in goroutine 31
        /mnt/vss/_work/1/s/cli/azd/pkg/async/task.go:140 +0x1c5
```

### PR DESCRIPTION
**Pull request by weikanglim**
Add a nil-check to handle when a python project is undetected. In this case, the python project will still be attempted to be built by the Oryx builder. This would likely result in a different, but friendlier error message and avoids the panic.

With this change, better error messaging is also provided when building a containerized project fails due to no Dockerfile and failed detection.

Fixes #2961

### CODE REVIEW

**Review by ellismg — APPROVED** [review]

File: cli/azd/pkg/project/framework_service_docker.go [hunk 1]
```diff
@@ -320,23 +320,25 @@ func (p *dockerProject) packBuild(
 				return nil, err
 			}
 
-			for _, dep := range prj.Dependencies {
-				if dep == appdetect.PyFastApi {
-					launch, err := appdetect.PyFastApiLaunch(prj.Path)
-					if err != nil {
-						return nil, err
+			if prj != nil {
```
Comment (vhvb1989) [inline cli/azd/pkg/project/framework_service_docker.go:323]: nit. Consider doing something about the indentation levels?  Probably break into smaller methods?

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/pkg/project/framework_service_docker.go (hunk 1 above)
Comment (wbreza) [inline cli/azd/pkg/project/framework_service_docker.go:323]: Consider inverting the `if` statement to reduce code nesting and also break some of these large `if` statements down into function calls to make the code a bit more readable and easy to digest.

**Review by wbreza — COMMENTED** [review]
Added some minor feedback.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 1 above)
Comment (weikanglim) [reply cli/azd/pkg/project/framework_service_docker.go]: I was trying to minimize the diff for the immediate change. However, given the overwhelming feedback, I've restored to the version that I had which includes the refactor.

**Pull request merged**